src/panels/lovelace/editor/hui-dialog-edit-card.ts

```diff
@@ -53,58 +72,131 @@ export class HuiDialogEditCard extends LitElement {
+          <paper-button
+            @click="${this._toggleEditor}"
+          >Toggle Editor</paper-button>
+          <paper-button
+            @click="${this._closeDialog}"
+          >Cancel</paper-button>
+          <paper-button
+            @click="${this._updateConfig}"'
+          >Save</paper-button>
         </div>
       </paper-dialog>
     `;
   }
 
-  private _handleYamlChanged(ev) {
-    this._previewEl.yaml = ev.detail.yaml;
+  private _handleYamlChanged(ev: YamlChangedEvent): void {
+    this._configValue = { format: "yaml", value: ev.detail.yaml };
+    this._updatePreview(this._configValue);
+  }
+
+  private _handleConfigChanged(value: LovelaceConfig): void {
+    if (this._configElement) {
+      this._configElement.setConfig(value);
+    }
+    this._configValue = { format: "js", value };
+    this._updatePreview(this._configValue);
+  }
+
+  private _updatePreview(value: ConfigValue) {
+    if (!this._previewEl) {
+      return;
+    }
+    this._previewEl.value = value;
   }
 
-  private _closeDialog() {
+  private _closeDialog(): void {
     this._dialog.close();
   }
 
-  private async _loadConfig() {
-    this._cardConfig = await getCardConfig(this.hass!, this._cardId!);
-    await this.updateComplete;
-    // This will center the dialog with the updated config
+  private _toggleEditor(): void {
+    if (this._editorToggle && this._configValue!.format === "js") {
+      this._currentConfigYaml = yaml.safeDump(this._configValue!.value);
+    } else if (this._configElement && this._configValue!.format === "yaml") {
+      this._configElement.setConfig(yaml.safeLoad(this._configValue!.value));
+    }
+    this._editorToggle = !this._editorToggle;
+  }
+
+  private async _loadConfig(): Promise<void> {
+    this._currentConfigYaml = await getCardConfig(this.hass!, this._cardId!);
+  }
+
+  private async _loadConfigElement(): Promise<void> {
+    const conf = yaml.safeLoad(this._currentConfigYaml);
+
+    const tag = conf.type.startsWith(CUSTOM_TYPE_PREFIX)
+      ? conf.type.substr(CUSTOM_TYPE_PREFIX.length)
+      : `hui-${conf.type}-card`;
+
+    const elClass = customElements.get(tag);
+    let configElement;
+    try {
+      configElement = await elClass.getConfigElement();
+      configElement.setConfig(conf);
+      configElement.hass = this.hass;
+      configElement.addEventListener("config-changed", (ev) =>
+        this._handleConfigChanged(ev.detail.config)
+      );
+      this._configValue = { format: "js", value: conf };
+      this._configElement = configElement;
+    } catch (err) {
+      this._configElement = null;
+    }
+
+    // This will center the dialog with the updated config Element
     fireEvent(this._dialog, "iron-resize");
   }
 
-  private async _updateConfig() {
-    const newCardConfig = this.shadowRoot!.querySelector("hui-yaml-editor")!
-      .yaml;
+  private async _updateConfig(): Promise<void> {
+    this._currentConfigYaml =
```

Comment:
This can remain a local variable, no need to be an instance?

Also, why does `currentConfigYaml` exists? I thought we have `this._configValue` to keep track of the current config.

Reply:
I thought about that but was unsure how the editor would react with `this._configValue.value` if the format was `JS`. Ill try it. It shouldn't ever get to the editor with it being `js` but I will test to make sure